webpack/hashed-folder-copy-plugin/src/HashedFolderCopyPlugin.ts

```diff
@@ -0,0 +1,377 @@
+    const { module, compilation, requireFolderOptions, resolver } = options;
+
+    // Map of asset names (to be prepended by the outputFolder) to asset objects
+    const assetsToAdd: Map<string, IAsset> = new Map<string, IAsset>();
+
+    for (const source of requireFolderOptions.sources) {
+      const { globsBase, globPatterns } = source;
+
+      let resolvedGlobsBase: string;
+      if (globsBase.startsWith('.')) {
+        // Does this look like a relative path?
+        resolvedGlobsBase = path.resolve(module.context, globsBase);
+      } else if (path.isAbsolute(globsBase)) {
+        // This is an absolute path
+        resolvedGlobsBase = globsBase;
+      } else {
+        // This looks like a NodeJS module path
+        let slashAfterPackageNameIndex: number;
+        if (globsBase.startsWith('@')) {
+          // The package name has a scope
+          slashAfterPackageNameIndex = globsBase.indexOf('/', globsBase.indexOf('/') + 1);
+        } else {
+          slashAfterPackageNameIndex = globsBase.indexOf('/');
+        }
+
+        let packageName: string;
+        let pathInsidePackage: string;
+        if (slashAfterPackageNameIndex === -1) {
+          packageName = globsBase;
+          pathInsidePackage = '';
+        } else {
+          packageName = globsBase.substr(0, slashAfterPackageNameIndex);
+          pathInsidePackage = globsBase.substr(slashAfterPackageNameIndex + 1);
+        }
+
+        let packagePath: string;
+        try {
+          const resolveResult: string = resolver.resolveSync(
+            {},
+            module.context,
+            `${packageName}/package.json`
+          );
+          packagePath = path.dirname(resolveResult);
+        } catch (e) {
+          const errorMessage: string = `Unable to resolve package "${packageName}"`;
+          compilation.errors.push(new Error(errorMessage));
+          return this._renderError(errorMessage);
+        }
+
+        resolvedGlobsBase = path.join(packagePath, pathInsidePackage);
+      }
+
+      for (const globPattern of globPatterns) {
+        const globResults: string[] = glob.sync(globPattern, { cwd: resolvedGlobsBase });
+        for (const globResult of globResults) {
+          const globResultFullPath: string = path.resolve(resolvedGlobsBase, globResult);
+
+          if (assetsToAdd.has(globResult)) {
+            const errorMessage: string = `Two files resolve to the same output path "${globResult}"`;
+            compilation.errors.push(new Error(errorMessage));
+            return this._renderError(errorMessage);
+          }
+
+          const assetContents: Buffer = FileSystem.readFileToBuffer(globResultFullPath);
+          assetsToAdd.set(globResult, { size: () => assetContents.byteLength, source: () => assetContents });
+          module.buildInfo.fileDependencies.add(globResultFullPath);
+        }
+      }
+    }
+
+    const hash: crypto.Hash = crypto.createHash('md5');
```

Comment:
I thought webpack has an api for hashing. Maybe on this.buildInfo https://cs.github.com/webpack/webpack/blob/0456807b72d914a507361d183c87380f740b40e1/lib/NormalModule.js#L646

Reply:
AKA if we have a hash available from the build as a whole, can we use that instead of generating a new one.

Reply:
Webpack does have a hash API, but I don't think it'll offer much beyond just using `crypto`.

> if we have a hash available from the build as a whole, can we use that instead of generating a new one.

The general use case for this plugin is for including a folder of static assets, which may not change between builds. If we use the whole compilation's hash, any change to the compilation will be a cache break on these assets, even if they didn't change.